### PR TITLE
Support for relational custom field conditions

### PR DESCRIPTION
**Pull request by bayangan1991**
Converted to poetry and blackfied the codebase in 32ac5075
**Added support for relational custom field conditions in** 5489be32
Rebuilt example app with tests b22007789be65397e13cab9f8e68c19aa83eb049

> PBI : WF-4247

# Relational Custom Field Conditions

In a workforce custom field, you can specify conditions on foreign keys to prevent rendering on the django forms. This conditions currently only work on fields directly on the model.

This PR introduces the ability for a condition to walk ORM relational model to apply conditions.

For example, with a custom field on an asset with the following model:
Client -> Property -> Asset

Currently, a field can only be conditionally shown based on the property it is assigned to:
`key=property value=1`
and for multiple properties
`key=property value=1,2,3...`

But, if you wanted a field to be visible for all properties on a client, you would have to list and maintain every property

With this PR you can walk the ORM model between ForeignKeys using `__`. So to add a condition for the field only be displayed on Client 1:
`key=property__client value=1`

This would limit the custom fields visibility for any assets assigned to properties on Client 1, without having to specify each property.

### CODE REVIEW

**Review by jarekwg — APPROVED** [review]
Code looks sound!
Great stuff poetrying/blacking it at the same time. Nice that you've done it in a separate commit. I think your original approach of submitting that as a separate PR altogether was even better, but this works too.

The current amount of tests in this repo is pretty dispappointing heh.. Would you mind adding a couple, just for this nested work you've added? (Also not fussed if we just merge this and add the tests in the WF repo instead..)

PS: shoot me your PyPI username, and I'll add you to the team so you can publish a new release of this package.

**Comment by bayangan1991** [comment]
Hey Jarek,
Tests have been added to workforce, will shoot you the link in private if you want to look over them.

**Comment by DamianHeard** [comment]
For what it is worth I would very much have preferred two PRs here.

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,5 @@ build/
 dist/
 django_user_defined_fields.egg-info
 example/database.db
+.idea/
+poetry.lock
```
[the remark this answers is not in the document]
Comment (bayangan1991) [reply .gitignore:9]: Hey, as per [the poetry documentation](https://python-poetry.org/docs/basic-usage/#commit-your-poetrylock-file-to-version-control) we don't commit the lock file for libraries.

File: userdefinedfields/factories.py [hunk 2]
```diff
@@ -1,14 +1,21 @@
-import factory
+try:
+    import factory
+except ImportError:
+    import sys
+
+    sys.stderr.write("Use of factories requires factory_boy\n")
+    sys.stderr.write("run pip install factory_boy\n")
```
[the remark this answers is not in the document]
Comment (bayangan1991) [reply userdefinedfields/factories.py:7]: This was more from a perspective of open source. I will remove the pip specific language though.

File: userdefinedfields/fields.py [hunk 3]
```diff
@@ -26,8 +26,19 @@ def _get_EXTRAFIELD_fieldlist(self, obj, field):
         filtered_fields = []
         for f in relevant_fields:
             for dc in f.displaycondition_set.all():
+                # Split the dc.key on '__' to create path to relevant field
+                # This still works for model fields as the traversal will
+                # only happen once
+                fields = dc.key.split("__")
```
[the remark this answers is not in the document]
Comment (DamianHeard) [reply userdefinedfields/fields.py:32]: do we want to putting some bounds on how many rels can be traversed?

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+(
+  /(
+    | migrations
+  )/
+  | models\.py
+  | urls\.py
+)
+'''
+
+[tool.flake8]
```
[the remark this answers is not in the document]
Comment (bayangan1991) [reply pyproject.toml]: Removed for now

**Review by DamianHeard — APPROVED** [review]
I'm going to tick this PR but suggest you probably want to bump the char limit on the key or the feature will have limited use.

File: userdefinedfields/fields.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bayangan1991) [reply userdefinedfields/fields.py:32]: It doesn't incur any extra SQL queries so I don't think there is any risk there. I've shown that in testing on WF

32 chars still gets you pretty far without it seeming limiting

File: userdefinedfields/fields.py [hunk 5]
```diff
@@ -26,8 +26,19 @@ def _get_EXTRAFIELD_fieldlist(self, obj, field):
...
+
                 try:
-                    dc_field = type(obj)._meta.get_field(dc.key)
+                    # Set the initial target model
+                    target_model = type(obj)
```
[the remark this answers is not in the document]
Comment (bayangan1991) [reply userdefinedfields/fields.py]: Applied

**Pull request merged**